airbyte-ci: refactor GradleTask to run in subprocess

alafanechere opened the pull request:
## What
We want to move away from containerized gradle execution in CI.
This PR makes `airbyte-ci` run gradle as a subprocess on the system, while keeping the `airbyte-ci` orchestration benefits.

## Example
I got[ a 🟢  run on destination-s3](https://github.com/airbytehq/airbyte/actions/runs/12892380100/job/35946483166) from this branch.

[Running a longer job on `source-postgres`, `destination-bigquery` and `destination-snowflake`.
](https://github.com/airbytehq/airbyte/actions/runs/12893107835)

## Question for reviewers ❓ 
- ~Which gradle options should we keep compared to the original containerized implementation~ : `--no-watch-fs` `--build-cache` and `--scan`
- ~Do we want live Gradle logging when executing the gradle task in airbyte-ci, I found it really noisy and piping gradle output to python then to console has some performance overhead.~ I'll try to provide live gradle output in a follow up PR but local testing has shown strange overhead when streaming gradle output to the console.

## How

### New GitHub Action for Java Environment Setup:
* [`.github/actions/install-java-environment/action.yml`](diffhunk://#diff-fcaa358e665313d155a26f49b81d810d36e7d464857a044320ab51fff7337d84R1-R32): Added a new action to install the Java environment, including Java and Gradle setup with configurable cache options.

### Updates to Build and Test Steps:
* [`.github/actions/run-airbyte-ci/action.yml`](diffhunk://#diff-ce432ff29f2547c1bfdfd6d8080fc9798498b6a7bcf12c1206c9a82592b3e461L118-R120): Integrated the new Java environment setup action into the CI workflow.
* [`airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/java_connectors.py`](diffhunk://#diff-15a5bb25c35f71f05ffdffb99af803c14b3d494b13b381c1670722a5dccd794bL63-R64): Updated the directory path for the distribution tar output.
* [`airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/java_connectors.py`](diffhunk://#diff-605f49180bb987341881bbe2256b59d6d27af6703e30bbbc279b78d3a99b528aL169-R169): Modified the test steps to use the new distribution directory path.

### Refactoring Gradle Task Handling:
* `airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py`: 
  - Removed the use of container-based Gradle task execution in favor of running Gradle commands in subprocesses.
  - Introduced a context manager to handle Gradle environment setup and teardown.
  - Simplified the Gradle task options and removed unnecessary properties and methods. [[1]](diffhunk://#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057L4-R23) [[2]](diffhunk://#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057L35-R161) [[3]](diffhunk://#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057L247-R179) [[4]](diffhunk://#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057L267-R194) [[5]](diffhunk://#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057L280-R214)

## DX benefits
- Faster runs - less overhead related to containerization
- More stable caching behavior - we get the usual S3 cache implementation
- Easier debugging, no need to dive into DockerInDocker complexity to investigate CI issues.
<img width="1143" alt="Screenshot 2025-01-21 at 18 36 35" src="https://github.com/user-attachments/assets/07abfe03-02ee-4c98-8699-4aa03c754092" />

alafanechere commented:
* **#52033** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/52033?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/52033?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py:
```diff
@@ -32,210 +36,129 @@ class GradleTask(Step, ABC):
-    CDK_MAVEN_METADATA_URL = (
-        "https://airbyte.mycloudrepo.io/public/repositories/airbyte-public-jars/io/airbyte/cdk/airbyte-cdk-core/maven-metadata.xml"
-    )
+    STATIC_GRADLE_OPTIONS = ("--build-cache", "--scan")
```
Which options shall we keep compared to the containerized implementation?

davinchia replied:
we should keep --no-daemon and --no-watch-fs since this is a single build so we don't need to incur the overhead of start the daemon/watching the fs that is mostly useful for systems that are long running.

postamar replied:
I don't know about --no-daemon, aren't we going to call gradle multiple times in the same VM now?

davinchia replied:
good question - I'm assuming we only run 1 connector test per CI runner and most devs will not invoke `airbyte-ci` locally, and will instead directly run gradle. Augustine, does that sounds right?

alafanechere replied:
> I'm assuming we only run 1 connector test per CI runner
Nope, the default airbyte-ci implementation is to run all modified connectors tests in the same runner, concurrently for python connectors, sequentially for java connectors as concurrent access to the gradle cache is not possible. I can try to revisit this logic later. Colocating execution on the same runner has benefits for python connectors for dagger caching. Following this change colocation for java connectors makes less sense and hurts DX.

In any case we do run multiple gradle command in a single connector pipeline execution: `distTar`, `unitTests`, `integrationTests`. 

**TLDR;**: I don't think we should add `--no-daemon` back. I'll re-add `--no-watch-fs`. Thanks for your inputs.

alafanechere commented on .github/actions/run-airbyte-ci/action.yml:
```diff
@@ -115,7 +115,9 @@ runs:
         ls -la airbyte-python-cdk || echo "No airbyte-python-cdk directory"
         ls -laL ../airbyte-python-cdk || echo "No airbyte-python-cdk symlink"
-
+    - name: Install Java Environment
```
For a followup: that would be cool to spare python connector only PR from installing the java env. We could rely on the `changes` output to check if any java connector was modified

davinchia commented on .github/actions/run-airbyte-ci/action.yml:
```diff
@@ -115,7 +115,9 @@ runs:
         ls -la airbyte-python-cdk || echo "No airbyte-python-cdk directory"
         ls -laL ../airbyte-python-cdk || echo "No airbyte-python-cdk symlink"
-
+    - name: Install Java Environment
```
clarification: does this run for all connector PR builds?

alafanechere replied:
Yes. But that could run conditionnally. I'm checking if its an easy fix ATM.
https://github.com/airbytehq/airbyte/pull/52033#discussion_r1924153466

alafanechere replied:
@davinchia I'm affraid there's not a straightforward quick way to conditionally run this.
We can't use rules changed files as nothing in file path determines that a connector is a java one.
We could run `airbyte-ci connectors --modified --language=java list` and check the output - but that could have its overhead too which might not be worth it given the speed of this step (~35sec).

davinchia replied:
agreed with your conclusion, 35 secs seems fine.

davinchia commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/java_connectors.py:
```diff
@@ -60,7 +60,8 @@ async def run_connector_build(context: ConnectorContext) -> StepResult:
         return build_connector_tar_result
-    dist_dir = await build_connector_tar_result.output.directory(dist_tar_directory_path(context))
+
+    dist_dir = await build_connector_tar_result.output.directory("build/distributions")
```
for understanding: why did this change? feels unrelated to the other changes, maybe I'm missing something.

alafanechere replied:
It changes because the original path related to an absolute in container path. the `build_connector_tar` step now returns a directory which is the connector directory

davinchia commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py:
```diff
@@ -32,210 +36,129 @@ class GradleTask(Step, ABC):
+        secrets_written = []
+        try:
+            # Check if the system has all the required commands in the path.
+            self.check_system_requirements()
```
making sure I understand: this is required to propagate the various gradle caching to the test subprocess? It's not automatically inherited?

alafanechere replied:
That's just a safety net to check that the dev / runner has all the required dependencies installed on their system. It will crash the pipeline with an explicit error instead of tryng to debug a less explicit gradle failure.

alafanechere replied:
I think your question related to the `gradle_environment` method. 
> is is required to propagate the various gradle caching to the test subprocess? It's not automatically inherited?

ATM the s3 creds can be passed either as a CLI option or as an env var and are then available as secrets in the context object. I prefer to not assume that the env var is already set and don't see an issue in overwriting it.

davinchia replied:
got it. I was reading too fast and assumed we are enforcing the presence of these S3 env vars - looks this is already optional, nice!

davinchia commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py:
```diff
@@ -32,210 +36,129 @@ class GradleTask(Step, ABC):
-            # When running locally, this dependency update is slower and less useful than within a CI runner. Skip it.
-            warm_dependency_cache_args = ["--dry-run"]
+            return stdout, stderr, process.returncode
+        finally:
```
if the started process hangs, how does this finally block trigger? 

this is probably my lack of python knowledge

alafanechere replied:
If it hangs it's never reached. Good point. I can add a timeout.

davinchia commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py:
```diff
@@ -277,14 +202,16 @@ async def _collect_test_results(self, gradle_container: Container) -> Optional[A
         if (
             test_results_dir_name_in_container
-            not in await gradle_container.directory(f"{self.context.connector.code_directory}/build").entries()
+            not in await self.dagger_client.host().directory(f"{self.context.connector.code_directory}/build").entries()
```
again for understanding, why did this change to dagger_client? Shouldn't the test output be retrieved from the started gradle test process?

alafanechere replied:
Dagger can access the host file system. To still handle the same type of objects we use for artifact uploads (dagger.File, dagger.Directory) I preferred to retrieve it the _dagger way_ instead of implementing something custom.

davinchia replied:
got it. So dagger best practices is to access all the OS pieces using the dagger client?

alafanechere replied:
It's just a way to seamlessly handle file and directory from host and containers in the same way, so you can move/mount them around with the same primitives.

davinchia reviewed (commented):
@alafanechere this looks sane to me. I'm not really qualified to review the Dagger code or Python best practices, so I mostly focused on understanding those pieces. Everything else seems good. 

One comment on the python process handling, but non-blocking. 

I see CI works. Did you manage to test this locally?

alafanechere commented:
> @alafanechere this looks sane to me. I'm not really qualified to review the Dagger code or Python best practices, so I mostly focused on understanding those pieces. Everything else seems good.

👍 

> One comment on the python process handling, but non-blocking.

I'll address that tomorrow.

> I see CI works. Did you manage to test this locally?

I did, but it was pretty slow - my machine was dying. I'm going to try again tomorrow with a less busy env.

alafanechere commented:
> > Do we want live Gradle logging when executing the gradle task in airbyte-ci, I found it really noisy and piping gradle output to python then to console has some performance overhead.
> 
> I realised I didn't answer this. What is the overhead of doing so? I think the use case for this is being able to see logs as tests run in the CI. I imagine it's useful if trying to debug in CI, however I'm less sure of the value since I'm not sure how the tests outputs will look like once this change goes in.
> 
> If the dx is the same as the current dagger tests, I think value is low, since the output is only at the end when everything is run. Value is high if we are able to see things live in the GHA UI, though lower if perf overhead is large. I think we can poll various Java devs and fast follow in another PR since this PR by itself has lots of value.

We can show the output live in the GHA console, but when I tried locally I had the impression that the output streaming had overhead. I can definitely experiment on this on a follow up PR.

frifriSF59 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/gradle.py:
```diff
@@ -32,210 +36,129 @@ class GradleTask(Step, ABC):
+                process.terminate()
+                process.wait(timeout=5)  # Give it 5 seconds to terminate gracefully
+                try:
+                    process.kill()  # Force kill if still running
```
Should this specific scenario be logged somehow? We might want to know that it happened?

postamar commented on .github/actions/install-java-environment/action.yml:
```diff
@@ -0,0 +1,32 @@
+        cache-write-only: ${{ inputs.gradle_cache_write_only }}
+    - name: Install system tools (jq, xargs)
+      shell: bash
+      run: sudo apt-get -y install jq findutils
```
Is this step really required? Have you tried without? jq is no longer needed (IIRC) and xargs should be present on the system (which is now a full fledged ubuntu and not some bare bones alpine distro)

alafanechere replied:
I removed it and it still works. Thanks for the suggestion

alafanechere commented:
> This seems like something worth doing! I'll note that there's more followup work in the form of ripping out all the stuff we did to support dind (wasn't /tmp mounted as a volume at some point?)

dind is still used for some python connector testing use cases which rely on docker-compose files.

alafanechere commented:
@postamar @davinchia @frifriSF59 thanks for your comments. Here are my latest changes following your suggestions.
- [Implement a timeout to avoid hanging gradle tasks.](https://github.com/airbytehq/airbyte/pull/52033/files#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057R144) I've set it to 2 hours by default. This can be customized for distTar / unit / integration tests by tweaking the value on the GradleTask subclasses.
- [I added the `--no-watch-fs` option](https://github.com/airbytehq/airbyte/pull/52033/files#diff-822a5e53d898c9dc906b45015dee19c141b79c97cdbe0f76bd9e77c02ee15057R52). I decided to not re-add the `--no-daemon` option because a single pipeline can run multiple gradle tasks.
- I added extra logging + made some readability improvement changes in `GradleTask`.
- Fixed airbyte-ci tests to get 🟢 CI.

I opened a followup PR for CI orchestration optimization: to leverage matrix strategies to run 1 java connector per runner and only install the java environment when needed (when testing java connectors). It's still a wip.

davinchia reviewed (approved):
:shipit: 

I'd be curious what are the pieces @postamar as no longer needed and can now be removed. However definitely not blocking.

postamar commented:
Apparently docker-in-docker is required for some python connectors so it can't be removed. This was by far the biggest piece. I'll take a look and see what odds and ends I can remove but they're fairly inconsequential in comparison.
